[MHV-47590] Update DeleteDraft Navigation

vsaleem opened the pull request:
## Summary
- _Resolved issue where 'Delete draft' button was not functional on unsaved messages._
- _New expected behavior: User is redirected to previous folder user was in after a unsaved draft message is deleted._

Several logic conditions have been updated, such as:
- _When a user clicks the 'Delete draft' button on Start a new message screen without making any input changes, an unsaved message will remain unsaved/deleted and the user is redirected to the previous location where the user started the new message._
- _When a user edits a partially composed message and then clicks the 'Delete draft' button on the 'Start a new message' screen, the user will be presented with a new modal, 'Draft this draft?' modal, and the user is given the option to continue editing the message or continue deleting the message. THEN, when a user confirms the message to be deleted, they are redirected to the previous location where the user started the new message._
- _This modal only displays on a newly composed type of message as unsaved and is not part of a conversation.
- _When message is saved and a user clicks the 'Delete draft' button on 'Start a new message' or Reply screen, the user will see the Delete Draft Modal, "Are you sure you want to delete this draft?".  This message warns the user that deleting messages from an existing conversation are permanent _
- _Existing logic around the RouteNavigatonGuard needed to be updated to not interfer with the 'Delete draft's modal logic.  The changes were meant to prevent unexpected behavior to the user, such as unexpected modal pop-ups or user being blocked from navigating away, and to offer an additional method to navigate away from a draft.  The delete button functionality is consistent on all SM pages where the 'Delete draft' button is present.

## Related issue(s)
### [MHV-47590](https://jira.devops.va.gov/browse/MHV-47590) - Update Delete Draft functionality on 'Start a new message page' ###

> 
> User Story: As a Secure Messaging user, I want to compose a message on the 'Start a new message' screen and have the option to delete an unsaved message and when I delete the message I want confirm my choice, so that I can be aware what happens to my message and can navigate away.
> 
> Acceptance Criteria:
> AC1 On 'Start a new message' screen, when the user makes no input changes to form, and when the user clicks 'Delete draft', the user can navigate away to it's respective folder
> AC2  On 'Start a new message' screen, when the user makes input changes to form, and when the user clicks 'Delete draft', the user receives 'Delete this draft?' modal. AND user can navigate away if action is confirmed and the unsaved message is deleted
> AC3 On 'Start a new message' screen, when the user makes input changes to form AND saves the message, and when the user clicks 'Delete draft', the user receives 'Are you sure you want to delete this draft?' modal. AND if action is confirmed the message will be deleted and the user can navigate away back to the respective folder
> AC4  'Start a new message' screen, when the user makes input changes to form and saves the message, and when the user clicks 'Delete draft'
> 

## Testing done
- Updated cypress tests: After deleting a message, the user is redirected to the initiating screen where the user began the message.
- Manually tested the functionality to see if correct delete modals appear and if navigation occurred
- Unit tests PENDING: adding test to accommodate unsaved draft modal

## Screenshots

|         | Before | After |
| ------- | ------ | ----- |

| Mobile  |N/A|Unsaved 'Start a new message' draft:<img width="345" alt="image" src="https://github.com/department-of-veterans-affairs/vets-website/assets/59583325/e509547a-308f-4b5f-b0e0-faa20da74bf2"> Saved messages draft:<img width="360" alt="Screenshot 2023-10-25 at 8 39 38 AM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/59583325/b8aab90c-7e8a-4371-8e4b-c916f0877aaa">|

| Desktop |N/A|'Start a new message' draft:<img width="492" alt="image" src="https://github.com/department-of-veterans-affairs/vets-website/assets/59583325/8681e881-fcc7-4d26-8ca0-73b1d4fa99dd">Saved messages draft:<img width="654" alt="Screenshot 2023-10-25 at 8 39 28 AM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/59583325/89916ead-6a73-4468-a9c8-0ee45fa27075">|

## What areas of the site does it impact?
- This code was created for Va.gov - Secure Messaging application.
*Va.gov - Secure Messaging*

## Acceptance criteria

- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [ ] Did you login to a local build and verify all authenticated routes work as expected with a test user

### :warning: Team Sites (only applies to modifications made to the VA.gov header) :warning:

- [ ] The vets-website header does not contain any web-components
- [ ] I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#local-dev) to test the injected header scenario
- [ ] I reached out in the `#sitewide-public-websites` Slack channel for questions

## Requested Feedback

Commit:
[MHV-47590] Update DeleteDraft Navigation

- user is redirected to previous folder user was in after a unsaved compose/reply draft message is deleted

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
[MHV-47590] Removed former unable to save alert; no longer needed

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge from 'main'

Commit:
[MHV-47590] Updated Delete Draft modal

- added constants for an unsaved new draft / alt. modal

- added props and updated scenarios to populate Delete Draft modal

- Delete Modal appears in Inbox folder

Commit:
[MHV-47590] Cleanup Navigation & DeleteModal Navigation

- Removed NavigationError alert from ReplyForm (this issue will be revisited in the future.)

- Updated navigation calls to function with Delete Draft modal

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
[MHV-47590] Removing cy test for 45118 update

- This updated needed to be removed and will be replanned for a future sprint.  Not totally removing the tests as it will be reintroduced at a later time.

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/components/ComposeForm/ComposeForm.jsx:
```diff
@@ -110,18 +125,41 @@ const ComposeForm = props => {
+            ErrorMessages.Navigation.UNABLE_TO_SAVE_DRAFT_ATTACHMENT_ERROR,
+          );
+          updateModalVisible(false);
+        }
```
we may want to cover this logic in a unit test

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/components/Draft/DeleteDraft.jsx:
```diff
@@ -53,15 +81,30 @@ const DeleteDraft = props => {
+            navigateToFolderByFolderId(
+              activeFolder ? activeFolder.folderId : DefaultFolders.INBOX.id,
+              history,
+            );
```
we may be missing test coverage for this block

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/tests/e2e/secure-messaging-navigate-away-from-replying-message.cypress.spec.js:
```diff
@@ -6,7 +6,7 @@ describe('verify navigate away pop-up message', () => {
   const site = new SecureMessagingSite();
 
-  it('navigate away by click on the inside link', () => {
+  it.skip('navigate away by click on the inside link', () => {
```
are we able to have this resolved with test engineers prior to merging to avoid skipping?

alex-morgun-1 reviewed (changes requested):
Vic, great work. Overall looks good. Can you look if we can add unit test coverage on modified areas of code and see if test engineers can assist with that test to avoid skipping it

fazilqa committed:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

KyleCardwell reviewed (approved):
Great job Vic, looks good!

fazilqa committed:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Asterix22 committed:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Asterix22 committed:
adding delete confirmation navigation cypress

Asterix22 committed:
more tests

Commit:
[MHV-47590] Added dispatch successful delete alert after 'deleting' a unsaved draft

Asterix22 committed:
add delete confirmation message verification

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Asterix22 committed:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
Merge branch 'main' into 47590-Delete-Draft-Modal-Update-mhv

Commit:
[MHV-47590] Cy test update

Commit:
[MHV-47590] Removed unused prop

vsaleem commented:
> Vic, great work. Overall looks good. Can you look if we can add unit test coverage on modified areas of code and see if test engineers can assist with that test to avoid skipping it

@oleksii-morgun-va can you take another look at this review?  I'm going to handle the unit tests for this ticket in a separate story --> https://jira.devops.va.gov/browse/MHV-51084

Commit:
[MHV-47590] Updated prop to delete unsaved, in progress reply draft message

- userSaved prop was removed which was used for this action; added new prop inProgressReplyDraft to check changes to messageBody before saving.

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/util/constants.js:
```diff
@@ -200,6 +205,8 @@ export const Prompts = {
     DELETE_DRAFT_CONFIRM: 'Are you sure you want to delete this draft?',
     DELETE_DRAFT_CONFIRM_NOTE:
       "Drafts are permanently deleted and this action can't be undone. \n\n Deleting a draft won't affect other messages in this conversation.",
+    DELETE_NEW_DRAFT_TITLE: 'Draft this draft?',
```
```suggestion
    DELETE_NEW_DRAFT_TITLE: 'Delete this draft?',
```

Commit:
[MHV-47590] Cleanup text change

rmessina1010 reviewed (approved):
LGTM
